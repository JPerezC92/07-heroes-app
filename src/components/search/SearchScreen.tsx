import React, { FormEventHandler } from "react";
import { heroes } from "../../data/heroes";
import HeroCard from "../heroes/HeroCard";
import useForm from "./useForm";

interface Props {}

const SearchScreen = (props: Props) => {
  const heroesFiltered = heroes;

  const { values, handleInputChange } = useForm({ searchClue: "" });

  const handleSearch: FormEventHandler<HTMLFormElement> = (e) => {
    e.preventDefault();
    console.log(values);
  };

  return (
    <div>
      <h1>SearchScreen</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Search</h4>

          <form onSubmit={handleSearch} autoComplete="off">
            <input
              type="text"
              placeholder="Search your hero"
              className="form-control"
              name="searchClue"
              value={values.searchClue}
              onChange={handleInputChange}
            />

            <button
              type="submit"
              className="btn mt-1 btn-block btn-outline-primary"
            >
              Search...
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {heroesFiltered.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </div>
  );
};

export default SearchScreen;
