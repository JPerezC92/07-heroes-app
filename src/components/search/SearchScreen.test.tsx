import { mount } from "enzyme";
import { MemoryRouter, Route, RouteComponentProps } from "react-router-dom";
import { routeComponentPropsFixture } from "../../test/fixtures/routeComponentPropsFixture";
import SearchScreen from "./SearchScreen";

describe("Test on <SearchScreen />", () => {
  test("should render correctly", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert-info").text().trim()).toBe("Search a hero");
  });

  test("should render Batman info and input text field with the queryString parameter", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.contains("Bruce Wayne")).toBeTruthy();
    expect(wrapper.find("input").props().value).toBe("batman");
    expect(wrapper).toMatchSnapshot();
  });

  test("should display error if the search fails", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=gomamon"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find(".alert-danger").text().trim()).toBe(
      "There is not a hero with gomamon"
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("should call push from history", () => {
    const searchScreenProps: RouteComponentProps = {
      ...routeComponentPropsFixture,
      history: { ...routeComponentPropsFixture.history, push: jest.fn() },
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route
          path="/search"
          component={(props: unknown) => (
            <SearchScreen {...props} {...searchScreenProps} />
          )}
        />
      </MemoryRouter>
    );

    wrapper.find("input").simulate("change", {
      target: { name: "searchText", value: "batman" },
    });
    wrapper.find("form").simulate("submit");

    expect(searchScreenProps.history.push).toHaveBeenCalledWith("?q=batman");
  });
});
