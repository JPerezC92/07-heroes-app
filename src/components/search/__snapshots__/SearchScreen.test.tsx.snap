// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <SearchScreen /> should display error if the search fails 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <form
        autoComplete="off"
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search your hero"
          type="text"
          value="gomamon"
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is not a hero with 
        gomamon
      </div>
    </div>
  </div>
</div>
`;

exports[`Test on <SearchScreen /> should render Batman info and input text field with the queryString parameter 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <form
        autoComplete="off"
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card"
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test on <SearchScreen /> should render correctly 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <form
        autoComplete="off"
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search your hero"
          type="text"
          value=""
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
